compiler/src/iree/compiler/Codegen/Common/GPU/GPULowerCoalescedDMAToGlobalLoads.cpp

```diff
@@ -0,0 +1,362 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cstdint>
+#include <numeric>
+#include <optional>
+#include "iree/compiler/Codegen/Common/GPU/Passes.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/DebugLog.h"
+#include "mlir/Dialect/AMDGPU/IR/AMDGPUDialect.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Arith/Utils/Utils.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/MemRef/Utils/MemRefUtils.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Utils/Utils.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/Dialect/Vector/Transforms/VectorTransforms.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-codegen-gpu-lower-coalesced-dma-to-global-loads"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_GPULOWERCOALESCEDDMATOGLOBALLOADSPASS
+#include "iree/compiler/Codegen/Common/GPU/Passes.h.inc"
+
+// Verifies that the mapping attributes match the expected pattern:
+// [#gpu.warp<linear_dim_1>, #gpu.warp<linear_dim_0>].
+static LogicalResult verifyWarpMapping(scf::ForallOp forallOp) {
+  std::optional<ArrayAttr> mappingAttr = forallOp.getMapping();
+  if (!mappingAttr) {
+    LDBG() << "No mapping attribute found";
+    return failure();
+  }
+  if (mappingAttr->size() != 2) {
+    LDBG() << "Mapping attribute size is " << mappingAttr->size()
+           << ", expected 2";
+    return failure();
+  }
+
+  auto warp0 = dyn_cast<gpu::GPUWarpMappingAttr>((*mappingAttr)[0]);
+  auto warp1 = dyn_cast<gpu::GPUWarpMappingAttr>((*mappingAttr)[1]);
+  if (!warp0 || !warp1) {
+    LDBG() << "  - Mapping attributes are not GPUWarpMappingAttr";
+    return failure();
+  }
+
+  if (warp0.getWarp() != gpu::MappingId::LinearDim1 ||
+      warp1.getWarp() != gpu::MappingId::LinearDim0) {
+    LDBG() << "  - Mapping mismatch (expected linear_dim_1, linear_dim_0)";
+    return failure();
+  }
+
+  return success();
+}
+
+// Verifies that destination memref is contiguous and source/target memory
+// address spaces are compatible.
+static LogicalResult verifyMemoryLayout(IREE::GPU::CoalescedGatherDMAOp dmaOp) {
+  // Check that destination memref is contiguous.
+  auto destMemRefType = cast<MemRefType>(dmaOp.getInit().getType());
+  LDBG() << "  - Destination type: " << destMemRefType;
+
+  if (!destMemRefType.getLayout().isIdentity()) {
```

Comment:
That condition's too strong - you could have LDS memrefs with a built-in offset.

I think you want `areTrailingDimsContiguous`

Reply:
... Ok, yeah, you want `areTrailingDimsContiguous(1)` probably, to save on checks